reference/src/layout/enums.md

```diff
@@ -351,90 +351,72 @@ enum Enum1<T> {
-interaction with uninhabited types.
+**Single variant data-carrying*** enums without a `repr()` annotation have the
+same layout as the variant field. **Single variant fieldless** enums have the
+same layout as a unit struct.
 
-As presently implemented, the compiler will use the same layout for
-structs and for single variant enums (as long as they do not have a
-`#[repr]` annotation that overrides that choice). So, for example, the
-struct `SomeStruct` and the enum `SomeEnum` would have an equivalent
-layout ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3697ac684c3d021892694956df957653)):
+Here:
 
 ```rust
-struct SomeStruct;
-enum SomeEnum {
-  SomeVariant,
-}
-
+# use std::mem::{size_of, align_of};
+struct UnitStruct;
+enum SingleVariantFieldless { FieldlessVariant }
 # fn main() {
-# use std::mem;
-let x = SomeStruct;
-let y = SomeEnum::SomeVariant;
-assert_eq!(
-    mem::size_of_val(&x), 
-    mem::size_of_val(&y),
-    "types have different sizes",
-);
-println!("types both have size {}", std::mem::size_of_val(&x));
+assert_eq!(size_of::<SingleVariantFieldless>(), size_of::<UnitStruct>());
+assert_eq!(align_of::<SingleVariantFieldless>(), align_of::<UnitStruct>());
+// assert_eq!(call_abi_of::<SingleVariantFieldless>(), call_abi_of::<UnitStruct>());
+// assert_eq!(niches_of::<SingleVariantFieldless>(), niches_of::<UnitStruct>());
 # }
 ```
 
-Similarly, the struct `SomeStruct` and the enum `SomeVariant` in this
-example would also be equivalent in their layout
-([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=924724764419f846c788a8763da45992)):
+the single-variant fieldless enum `SingleVariantFieldless` has the same layout
+as `UnitStruct`.
+
+Here: 
 
 ```rust
+# use std::mem::{size_of, align_of};
 struct SomeStruct { x: u32 }
-enum SomeEnum {
-  SomeVariant { x: u32 },
+enum SingleVariantDataCarrying {
+  DataCarryingVariant(SomeStruct),
 }
-
 # fn main() {
-# use std::mem;
-let x = SomeStruct { x: 22 };
-let y = SomeEnum::SomeVariant { x: 22 };
-assert_eq!(
-    mem::size_of_val(&x), 
-    mem::size_of_val(&y),
-    "types have different sizes",
-);
-println!("types both have size {}", mem::size_of_val(&x));
-}
+# assert_eq!(size_of::<SingleVariantDataCarrying>(), size_of::<SomeStruct>());
+# assert_eq!(align_of::<SingleVariantDataCarrying>(), align_of::<SomeStruct>());
+# }
 ```
 
-In fact, the compiler will use this optimized layout even for enums
-that define multiple variants, as long as all but one of the variants
-is uninhabited
-([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3cc1484c5b91097f3dc2015b7c207a0e)):
+the single-variant data-carrying enum `SingleVariantDataCarrying` has the same
+layout as `SomeStruct`.
+
+### Layout of multi-variant enums with one inhabited variant
+
+The layout of **multi-variant** enums with **one inhabited variant** is the same
```

Comment:
Have we defined "inhabited" sufficiently precisely for this?

I am a bit worried about the details of this definition.

Reply:
There is no definition of uninhabited or inhabited types anywhere in the reference. 

What worries you about this definition ? If only one variant is inhabited, all other variants are uninhabited, and there can be no value of the enum that contains any of those variants, according to all definitions of uninhabited types that I can imagine.

Reply:
There are at least three different kinds of "inhabited":
* Type tas a call ABI other than `Uninhabited`.
* The type's validity invariant is satisfiable.
* The type's safety invariant is satisfiable.

So, the term is ambiguous without further clarification.

Reply:
Are there types for which "The type's validity invariant is satisfiable" and the call ABI is `Uninhabited` ?

Reply:
Good question! I don't know. `&!` might be such a type, depending on how the validity invariant discussions turn out.

Declaring that these two notions (the first two in my list) are equal would put a restriction on the validity invariant discussion: we would have to make sure that satisfiability is trivially computable and can be put into the rustc layout code. I think this would be a nice property to have, but I wouldn't want to assert now that we will end up there.

Reply:
We defined that `&T where T: Sized` has the same alignment and size as an `usize`. _I think_ this means that the call ABI of `&T where T: Sized` is `Scalar`. Is this correct? If so, for `&!` to have an `Uninhabited` call ABI `!` must not be `Sized`.

Reply:
So we might need to go back to the layout of pointers and references and require `where T: Sized + SomeFormOfInhabited`.

Reply:
> If so, if &! has an Uninhabited call ABI then ! cannot be Sized, which would mean that any type containing a ! is not Sized..

Not really -- if an `&!` is impossible to construct, it doesn't really matter which call ABI we ascribe to it.

Reply:
But this does indeed show that treating "uninhabited" as a call ABI might not be the best idea, at least in terms of teaching.

Reply:
> it doesn't really matter which call ABI we ascribe to it.

Then we don't have to give it Uninhabited as a call ABI, and we go back to my original question (Are there types for which "The type's validity invariant is satisfiable" and the call ABI is Uninhabited ?). 

---

We say in many places that `&T` and `*T` have the same call ABI when `T: Sized`, we would need to add exceptions for `&!` and `*!` if this is not the case. Also we do need to deal with types for which the call ABI of `&!` might matter, e.g., what's the call ABI of `Option<&!>` ? Ideally we want it to be compatible with `*!`, but I don't see how to make that happen if `&!` has a different call ABI than `*!`.

Reply:
If we stop considering "uninhabited" a call ABI, what happens is that "layout" gets one more property: "inhabited".

So the first item of my list changes to "The type is considered inhabited in its layout".

Reply:
> what's the call ABI of Option<&!>

Oh gosh, I have no idea.^^ Probably if we really made `&!` uninhabited that would become a ZST... ouch.

I am not sure what the right framework here is. But people have been bringing up that being able to optimize away methods taking `&!` (because they are dead code) could be very beneficial. That's the main motivation for caring about validity of the memory "behind" the reference in https://github.com/rust-lang/unsafe-code-guidelines/issues/77.

Reply:
Ok, so we really do need to work out the layout of uninhabited types before we do anything else here.

I do think that "inhabited", whatever that means, will need to hold here for this guarantee. Whether `&!` is inhabited or not, doesn't change anything here AFAICT.

Reply:
Popping back up, I don't think we should talk about "(un)inhabited" types in these documents before we got those definitions set. So I am in favor of dropping this subsection.